### ISSUE TITLE
Recommendation: Remove multiple decoding steps and redundant tx model   

### ISSUE THREAD

**Issue by thec00n** [omisego/plasma-contracts#458]
## Issue Type 

```
[ ] bug report
[ X ] feature request
```

## Current Behavior

In various ExitGame flows transactions are decoded multiple times which leads to unnecessary 
gas costs. Further there are structs decoded from RLP structures that are not strictly necessary. The audit efforts have revealed that it is important to the security of the tx signing and Merkle proofs that there is a 1:1 relationship between encoded and decoded tx. Explicit and well defined decoding routines are going to be adopted and it's clear that they will enforce a model both in terms of encoding as well as its structure. 

The following example is from the PaymentChallengeStandardExit lib and should demonstrate how in the same ExitGame flow decoding is performed multiple times.

An exiting tx is decoded in the following places:

https://github.com/omisego/plasma-contracts/blob/1c7cd867b1353ae72c825a6d15b2d382689f30ad/plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeStandardExit.sol#L111-L114

https://github.com/omisego/plasma-contracts/blob/1c7cd867b1353ae72c825a6d15b2d382689f30ad/plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeStandardExit.sol#L147-L150

https://github.com/omisego/plasma-contracts/blob/1c7cd867b1353ae72c825a6d15b2d382689f30ad/plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeStandardExit.sol#L161-L168

A challenging tx is decoded in the following places:

https://github.com/omisego/plasma-contracts/blob/1c7cd867b1353ae72c825a6d15b2d382689f30ad/plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeStandardExit.sol#L128

https://github.com/omisego/plasma-contracts/blob/1c7cd867b1353ae72c825a6d15b2d382689f30ad/plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeStandardExit.sol#L151

https://github.com/omisego/plasma-contracts/blob/1c7cd867b1353ae72c825a6d15b2d382689f30ad/plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeStandardExit.sol#L161-L168

## Suggested Fix

**Remove unnecessary data model**

`WireTransaction` is a subset of `PaymentTransactionModel` and it's not strictly needed for the current tx type. It is recommend to remove it. 

https://github.com/omisego/plasma-contracts/blob/1c7cd867b1353ae72c825a6d15b2d382689f30ad/plasma_framework/contracts/src/transactions/WireTransaction.sol#L11

**Decode a tx once and pass it down stream**

Most Exit Game flows do preprocessing for specific function arguments before they process the structure down stream. It is recommended to decode `PaymentTransactionModel` once and then make it available throughout the Exit Game flow. 

**Be more explicit with PaymentTransactionModel checks**

`PaymentTransactionModel` does allow between 0 to 4 inputs and 0 to 4 outputs. This model is in fact a generalization of a standard tx and a deposit tx that have different structures. A deposit tx has 0 inputs and 1 output while a standard tx has between 1-4 inputs and 1-4 outputs. While it makes sense to have a generalized model it should still be checked during the verification process which type of tx is processed based UTXO pos.

**Comment by thec00n** [comment]
Did not see the issue https://github.com/omisego/plasma-contracts/issues/456. A lot is covered here already. I would prefer to decode txs once and then do specific check on the structure later on to distinguish between the standard and deposit tx.

**Comment by kevsul** [comment]
> WireTransaction is a subset of PaymentTransactionModel and it's not strictly needed for the current tx type. It is recommend to remove it.

WireTransaction is more of a "base class" Transaction. We can change the name to something like `BaseTransaction` or `GenericTransaction` to make it clearer.

We do need something like this though, because there are times when you don't know what type of transaction you're dealing with and you need to be able to partially decode it to get the `txType` and an `outputType` to verify the spending condition. 
Starting or challenging an IFE are examples of when this occurs.

You're right that we decode transactions many times unnecessarily. I suggest modifying WireTransaction to provide a minimal decoding method that just does enough to get the types (as well as doing some common checks)and returns the intermediate decoded tx so that concrete transaction types can build on it without repeating the decoding.

**Comment by thec00n** [comment]
`WireTransaction is more of a "base class" Transaction. We can change the name to something like BaseTransaction or GenericTransaction to make it clearer.`
Generally if that is the intention then that makes a lot of sense.

`We do need something like this though, because there are times when you don't know what type of transaction you're dealing with and you need to be able to partially decode it to get the txType and an outputType to verify the spending condition.`

`WireTransaction` does not have expose the `txType` only the `outputType`. It be helpful to understand what the future plan is is with this type of abstraction. If there is` BaseTransaction` type then what are attributes that all tx share so the contracts can some type of preprocessing. I assume that a base contract must have access to both  `txType` only the `outputType`.

**Comment by kevsul** [comment]
`WireTransaction` does currently have `getTransactionType()` to expose `txType`

My idea is to have a `GenericTransaction` that can decode any rlp encoded tx into:
```
    struct BaseTransaction {
        uint256 txType;
        RLPReader.RLPItem[] inputs;
        RLPReader.RLPItem[] outputs;
        RLPReader.RLPItem txData;
    }
```

and it would also be able to decode an output into a BaseOutput (like current behaviour):
```
    struct BaseOutput {
        uint256 outputType;
        bytes20 outputGuard;
        address token;
        uint256 amount;
    }
``` 

We can then pass around a `BaseTransaction` instead of the rlp encoded transaction, which will save us from having to decode it multiple times. Concrete transaction types can do extra decoding of the `BaseTransaction` when necessary.
 
This is probably easier to explain with a POC, I'll get one together

**Comment by kevsul** [comment]
Fixed in #502
